src/ledger/LedgerTxn.h

```diff
@@ -635,13 +635,20 @@ class LedgerTxn : public AbstractLedgerTxn
     std::unique_ptr<Impl> const& getImpl() const;
 
   public:
+    // WARNING: use useTransaction flag with caution. It does not start a SQL
+    // transaction, which uses the strongest SERIALIZABLE level isolation.
+    // Therefore, if you have concurrent transactions, you are risking getting
+    // inconsistent view of the database. Only use this mode for read-only
+    // transactions with no concurrent writers present.
     explicit LedgerTxn(AbstractLedgerTxnParent& parent,
-                       bool shouldUpdateLastModified = true);
-    explicit LedgerTxn(LedgerTxn& parent, bool shouldUpdateLastModified = true);
+                       bool shouldUpdateLastModified = true,
+                       bool useTransaction = true);
```

Comment:
I'm not sure that this interface satisfies the principle of easy to use correctly, hard to use incorrectly. Let me give you an example. At a certain moment, a maintainer believes that a certain `LedgerTxn` actually represents a read-only transaction with no concurrent writers. The maintainer sets `useTransaction = false`. Later the code evolves such that the parent of that `LedgerTxn` is not guaranteed to be a true root. The code still compiles, but there is a performance regression.

It would be great if it was only possible to compile the code in cases where the flag actually made sense.

Reply:
Discussed with Jon that it is pretty annoying/invasive to make this change with the current structure of LedgerTxn roots. We can make the fix later, so I opened an issue to track this: https://github.com/stellar/stellar-core/issues/3197